### ISSUE TITLE
Better handling of timeout error from GitHub with an invalid response body

### ISSUE THREAD

**Issue by autarch** [graphql-rust/graphql-client#467]
For a while I was trying to figure out why sometimes I'd get an empty response body for GitHub GraphQL requests. Eventually I ended up replacing the `post_graphql` function with my own version that dumped more info from the response. I found out that GitHub sometimes responds with a JSON body like this:

```json
{"message": "We couldn't respond to your request in time. Sorry about that. Please try resubmitting your request and contact us if the problem persists."}
```

The current implementation of the graphql-client code makes this entirely invisible to the user of the library. I'm really not sure what the fix is here. I'm pretty sure that GitHub's response is just wrong. They should return a document with `errors` set to something. But they don't.

It would be really nice if this library would handle this case. I think ideally it'd return some sort of "invalid response format" error that contains the original body. As it stands, it seems like it just tries to parse the response and ends up creating an empty `graphql_client::Response` struct, which is really confusing.

**Comment by autarch** [comment]
I reported this to GitHub as well, since what they're doing is simply wrong. https://github.com/orgs/community/discussions/78614

**Comment by TieWay59** [comment]
I have a similar itch in debug&guard `post_graphql` because I need to check the response headers, so I rewrite the original function. Hello @tomhoule, do you have any better idea on designing this? 

FYI: In `reqwest::` crate `response.json()` would consume the response instance in this context so you have to clone other data before returning the body.

```rust
pub fn post_graphql_blocking<Q: GraphQLQuery, U: reqwest::IntoUrl + Clone>(
    client: &reqwest::blocking::Client,
    url: U,
    variables: Q::Variables,
    mut f: impl FnMut(&reqwest::header::HeaderMap) -> anyhow::Result<()>,
) -> Result<graphql_client::Response<Q::ResponseData>, reqwest::Error>
where
    Q::Variables: Window,
{
    // original impl

    // take response headers out
    let _ = f(response.headers());

    response.json()
}
```

**Comment by tomhoule** [comment]
Hmm yes, good point. I'm not sure what the best way out of this could be. The whole point of the included client is that it returns a typed response, so there isn't much to gain if it returns something from reqwest directly anymore. My personal preference would be removing the client, but I'm not sure how much that would break people's workflows.

**Comment by TieWay59** [comment]
@tomhoule 

I agree, in this case, it's better to let users write their posting logic. I write a lot of guarding checking code in `post_graphql_blocking` directly after I rewrite it in my package.

>  I'm not sure how much that would break people's workflows.

Much, because using `post_graphql_blocking` is the most simple example in the first page docs. We'll need another leading example because the generated structs are not so obvious at first glance.

**Comment by tomhoule** [comment]
Is there a way to split the body from the headers in reqwest? We could return (graphql_body, headers) from.

**Comment by TieWay59** [comment]
> Is there a way to split the body from the headers in reqwest? We could return (graphql_body, headers) from the graphql_client wrapper. It wouldn't solve the issue of non-graphql-compliant bodies though, we would have to have an extra variant somewhere for bodies that aren't valid graphql responses.

@tomhoule 

My solution is below, but it seems not the best:

```rust
fn get_headers_and_response_body_from_reqwest_result(
    reqwest_response: Result<reqwest::blocking::Response, reqwest::Error>,
) -> (reqwest::header::HeaderMap, String) {
    match reqwest_response {
        Ok(r) => {
            let header_map = r.headers().clone();
            let body = r.text().unwrap_or("respnse.text() failed".to_owned());
            (header_map, body)
        }
        Err(e) => {
            log::error!("reqwest_response is Err: {e:#?}");
            (
                reqwest::header::HeaderMap::new(),
                "reqwest_response is Err".to_owned(),
            )
        }
    }
}
```

**Comment by autarch** [comment]
> It wouldn't solve the issue of non-graphql-compliant bodies though, we would have to have an extra variant somewhere for bodies that aren't valid graphql responses.

I think the right way to do this would be to have `post_graphql` return a `Result`, rather than just a tuple of headers & body. Then an invalid GraphQL response is just an `Err`.

If you wanted to keep the old funcs for back-compat I think the new ones could be named `try_post_graphql` and `try_post_graphql_blocking`.

I think you'd need to parse the response into a `serde_json::Value`, check that it's an `Object`, _then_ check it's keys to make sure it has one of the two required keys and no extras, and _then_ finally turn that into the `Q::ResponseData` type. I just looked and there's [a `serde_json::from_value` fn](https://docs.rs/serde_json/latest/serde_json/value/fn.from_value.html) that will do this conversion for you.

**Comment by TieWay59** [comment]
Thank you @autarch, I've been in trouble lately, so I may need a few weeks to proceed with this issue. I agree with the `try_post_graphql*` naming. My solution above is too simple to align with the original proposal.
